### PR TITLE
[Form] Twig example of personalized embed collection

### PR DESCRIPTION
**Pull request by Th3Mouk**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | none |

### CODE REVIEW

File: book/forms.rst [hunk 1]
```diff
@@ -1574,6 +1574,15 @@ are 4 possible *parts* of a form that can be rendered:
 
 .. note::
 
+    The portion corresponding to ``collection`` type is ``entry``.
+    In the case of embed collections, for exemple,
```
Comment (javiereguiluz) [inline book/forms.rst]: `embed collections` ->  `embedded collections`

Comment (javiereguiluz) [inline book/forms.rst]: `for exemple` -> `for example`

File: book/forms.rst [hunk 2]
```diff
@@ -1574,6 +1574,15 @@ are 4 possible *parts* of a form that can be rendered:
...
+    in a form which you have a year which contains months, and months contains days.
+    You want to personnalize each collection render, where the form prefix is ``_my_form_year``,
```
Comment (javiereguiluz) [inline book/forms.rst]: `personnalize` ->  `personalize`

**Comment by javiereguiluz** [comment]
@Th3Mouk there are some minor typos to fix, but I like your proposal. Thanks for contributing to Symfony!

**Comment by Th3Mouk** [comment]
@javiereguiluz thanks for support !

File: book/forms.rst [hunk 3]
```diff
@@ -1568,6 +1568,15 @@ are 4 possible *parts* of a form that can be rendered:
 
 .. note::
 
+    The portion corresponding to ``collection`` type is ``entry``.
+    In the case of embedded collections, for example,
+    in a form which you have a year which contains months, and months contains days.
+    You want to personalize each collection render, where the form prefix is ``_my_form_year``,
+    you need to write two blocks, ``_my_form_year_months_entry_row`` to customize month render, and
+    ``_my_form_year_months_entry_days_entry_row`` for days inputs.
```
Comment (xabbuh) [inline book/forms.rst]: To be honest, I do not really understand this addition. To which `entry` do we refer here? And what do we understand by "a year which contains months, and months contains days"?

Comment (Th3Mouk) [inline book/forms.rst]: Its an example on multiple embedded collections, a form which contains a collectiontype, which contains himself an other collectiontype. If you have an idea for a better explanation, it would help.

Comment (xabbuh) [inline book/forms.rst]: Do you have the code of the example you have in mind? If I completely understand what you are thinking about I can hopefully suggest something that is (as I think) a bit easier to understand (maybe it's sufficient to provide some more details).

Comment (Th3Mouk) [inline book/forms.rst]:
```php

namespace AppBundle\Form\Type;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\CollectionType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class CalendrierExerciceType extends AbstractType
{
    /**
     * @param FormBuilderInterface $builder
     * @param array                $options
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('periodeSemaines', CollectionType::class, [
                'label' => false,
                'entry_type' => PeriodeSemaineType::class,
                'entry_options' => ['saison' => $options['saison']],
                'by_reference' => false,
                'allow_add' => true,
                'prototype_name' => '__periode__',
            ])
        ;
    }

    /**
     * @param OptionsResolver $resolver
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(array(
            'saison' => null,
            'data_class' => 'Entities\ElcaBundle\Entity\CalendrierExercice',
        ));
    }
}
```

```php

namespace AppBundle\Form\Type;

use Entities\ElcaBundle\Repository\PeriodeRepository;
use Symfony\Bridge\Doctrine\Form\Type\EntityType;
use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\CollectionType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class PeriodeSemaineType extends AbstractType
{
    /**
     * @param FormBuilderInterface $builder
     * @param array                $options
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('periode', EntityType::class, [
                'class' => 'Entities\ElcaBundle\Entity\Periode',
                'choice_label' => 'libelle',
                'query_builder' => function (PeriodeRepository $em) use ($options) {
                    return $em
                        ->createQueryBuilder('p')
                        ->join('p.saison', 's')
                        ->where('s.id = :id_saison')
                        ->setParameter('id_saison', $options['saison']->getId())
                    ;
                },
                'attr' => [
                    'data-select2-perso' => true,
                ],
            ])
            ->add('semaines', CollectionType::class, [
                'entry_type' => SemaineZoneType::class,
                'by_reference' => false,
                'allow_add' => true,
                'prototype_name' => '__semaine__',
            ])
        ;
    }

    /**
     * @param OptionsResolver $resolver
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(array(
            'data_class' => 'Entities\ElcaBundle\Entity\PeriodeSemaine',
            'saison' => null,
        ));
    }
}
```

```php

namespace AppBundle\Form\Type;

use Symfony\Bridge\Doctrine\Form\Type\EntityType;
use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\DateType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;

class SemaineZoneType extends AbstractType
{
    /**
     * @param FormBuilderInterface $builder
     * @param array                $options
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('startDate', DateType::class, [
                'label' => 'Date de début',
                'widget' => 'single_text',
                'format' => 'dd/MM/yyyy',
            ])
            ->add('endDate', DateType::class, [
                'label' => 'Date de fin',
                'widget' => 'single_text',
                'format' => 'dd/MM/yyyy',
            ])
            ->add('zones', EntityType::class, [
                'class' => 'Entities\ElcaBundle\Entity\Zone',
                'choice_label' => 'libelle',
                'multiple' => true,
                'required' => false,
                'attr' => [
                    'data-select2-perso' => true,
                ],
            ])
        ;
    }

    /**
     * @param OptionsResolver $resolver
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(array(
            'data_class' => 'Entities\ElcaBundle\Entity\Semaine',
        ));
    }
}
```

```twig
{% block _calendrier_exercice_periodeSemaines_entry_row %}
    <div class="row periode">
        <div class="col-xs-2">
        {{ form_row(form.periode) }}
        </div>
        <div class="col-xs-10">
        {{ form_errors(form.semaines) }}
        {{ form_label(form.semaines) }}<a href="#" data-add-week style="color: green;"> <i class="fa
        fa-plus-circle"></i></a>
        {{ form_widget(form.semaines) }}
        </div>
    </div>
{% endblock %}

{% block _calendrier_exercice_periodeSemaines_entry_semaines_entry_row %}
    <div class="row semaine">
        <div class="col-xs-4">
            {{ form_row(form.startDate) }}
        </div>
        <div class="col-xs-4">
            {{ form_row(form.endDate) }}
        </div>
        <div class="col-xs-4">
            {{ form_row(form.zones) }}
        </div>
    </div>
{% endblock %}
```

Comment (HeahDude) [inline book/forms.rst]: @Th3Mouk Those are nice examples! However I suggest minor things if they should go in the docs:
- utiliser des noms de classes et d'options en anglais "saison => season",
- actually `PeriodeSemaineType` may break
  - This is a good opportunity to show some option resolving as default for "saison" option is null but you call `$options['saison']->getId()`
    You should have in `configureOptions`
    ```php
    $resolver->setDefaults(array(
        'data_class' => 'Entities\ElcaBundle\Entity\PeriodeSemaine',
        'saison' => 'null',
        'query_builder' => function (Options $options) { // make it lazy
            $saison = $options['saison'];
            if (null  === $saison || ! ($saison instanceof Saison) {
                return null;
            }
            
            return function (PeriodeRepository $em) use ($saison) {
                return $em
                    ->createQueryBuilder('p')
                    ->join('p.saison', 's')
                    ->where('s.id = :id_saison')
                    ->setParameter('id_saison', $saison->getId())
                ;
            };
        },
    ));

    Btw `CalendrierExerciceType` only needs to have `$resolver-setDefined('saison')` as a "child" option it needs no default value, only to be passed to the child which will resolve it.

Comment (Th3Mouk) [inline book/forms.rst]: I don't know if the whole code must be added to documentation, but it's an example of what i've didn't find in the customization doc.

File: book/forms.rst [hunk 4]
```diff
@@ -1568,6 +1568,15 @@ are 4 possible *parts* of a form that can be rendered:
 
 .. note::
 
+    The portion corresponding to ``collection`` type is ``entry``.
```
Comment (HeahDude) [inline book/forms.rst]: @Th3Mouk I'm sorry I thought your example was for the doc and not about illustrating what you think is missing in the docs :)

In fact, I think there is a misunderstanding about the ``entry`` portion in your phrasing.

The portion corresponding to `collection` is `collection` (see [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Twig/Resources/views/Form/form_div_layout.html.twig#L26)) which can get overridden by default with the field name, in your example `periodeSemaine`:
```php
$builder->add('periodeSemaines', CollectionType::class, // ...
```

But you can define it using the `block_name` option (ref [doc](http://symfony.com/doc/current/reference/forms/types/form.html#block-name)).

`CollectionType` actually overrides the `block_name` value for its entries options with `entry` using a normalizer (see [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php#L85)).

So in your case (I presume):
```php
// Controller
$form = $this->createBuilder()
    ->add('calendrier_exercice', CalendrierExerciceType::class, array(
    // let you override 'form_row' with '_calendrier_exercice_row'
    // ...

// CalendrierExerciceType.php
$builder->add('periodeSemaines', CollectionType::class, array(
    // 'collection' => 'periodeSemaines'
    'entry_type' => PeriodeSemaineType::class,
    // 'form' => 'entry', overridden for all entries
    // ...

// PeriodeSemaineType
$builder->add('semaines', CollectionType::class, array(
    // 'collection' => '_semaines'
    'entry_type' => SemaineZoneType::class,
    // 'form' => 'entry'
    // ...
```
So you get:
 `form_collection_entry_collection_entry` default "cascade" block name which you can override with `_calendrier_exercice_periodeSemaines_entry_semaines_entry`

Note that you could have:
```php
// PeriodeSemaineType
$resolver->setDefault('block_name', 'periode_semaine');

// SemainZoneType
$resolver->setDefault('block_name', 'semaine');
```

and just do:
```twig
{# no underscore needed as 'block_name' is explicitly set #}
{% block periode_semaine_entry_row %}
{% block semaine_entry_row %}
```

Conclusion
========
`entry` will always override `block_name` option even defined in a form type if it's nested in a `CollectionType`.

And you're right that should be mentioned in the docs! ;)

Perhaps it should also be mentioned in the `CollectionType` reference as well. Could be a note near `entry_options` or in the introduction ?

**Comment by twify93** [comment]
I am totally going nuts with this collection type! I renamed my parent form to mainForm, and when I try to use `form` in the block it gives me `Variable "form" does not exist`:

```
{% set bookingSealNumbersForm = mainForm.bookingSealNumbers %}
{% form_theme bookingSealNumbersForm _self %}

{% block _mainForm_bookingSealNumbers_entry_row %}
    {{ form.count }}
    {#{{ form_label(form.sealNumber, null, {'label_attr': {'class': 'col-sm-2 control-label'} }) }}#}
    {#<div class="form-group {% if not form.sealNumber.vars.valid %}has-error{% endif %}">
        {{ form_label(form.sealNumber, null, {'label_attr': {'class': 'col-sm-2 control-label'} }) }}
        <div class="col-sm-4">
            {{ form_widget(form.sealNumber, { 'attr': { 'class': 'form-control' }}) }}
            {{ form_errors(form.sealNumber) }}
        </div>
    </div>#}
{% endblock %}
```

I am showing the comment lines too because I want you to see what I've tried, the local variable inside the block `form` doesn't exists and I've tried lots and lots of things, but no luck. I also tried to use the fully qualified names, e.g. `mainForm.bookingSealNumbers.sealNumber` with no luck, tried all sort of things...

**Comment by stof** [comment]
@twify93 your issue is that you use ``_self`` as form theme, but your current template does not extend anything. this means that the ``{% block _mainForm_bookingSealNumbers_entry_row %}`` is rendered as part of the normal template rendering too, not just when using the form theming.
Actually, this is why I always store form theming blocks in a separate template rather than in ``_self``. Mixing 2 different rendering contexts in the same template creates far too much opportunities to break things compared to having 2 templates

**Comment by twify93** [comment]
OK, so simple example which can work? I've moved the theme to another file, but now it's just ignoring it... It's like I am not getting the name of the row I want to override..

**Comment by stof** [comment]
Well, have you updated your ``{% form_theme %}`` tag to point to this other template rather than to ``_self`` after moving the theme ?

**Comment by twify93** [comment]
OK, so simple example which can work? I've moved the theme block to another file, but now it's just ignoring it... It's like I am not getting the name of the row I want to override..

`{% form_theme mainForm '...../sealNumberForm.html.twig' %}`

and `sealNumberForm.html.twig`:
```
{% block _main_form_bookingSealNumbers_entry_row %}
    {{ form.count }}
    {{ form_label(form.sealNumber, null, {'label_attr': {'class': 'col-sm-2 control-label'} }) }}
    <div class="form-group {% if not form.sealNumber.vars.valid %}has-error{% endif %}">
        {{ form_label(form.sealNumber, null, {'label_attr': {'class': 'col-sm-2 control-label'} }) }}
        <div class="col-sm-4">
            {{ form_widget(form.sealNumber, { 'attr': { 'class': 'form-control' }}) }}
            {{ form_errors(form.sealNumber) }}
        </div>
    </div>
{% endblock %}
```

I also tried with `{% block _booking_bookingSealNumbers_entry_row %}`, `{% block _main_form_bookingSealNumbers_entry_row %}`, _mainForm_bookingSealNumbers_entry_row %}` also tried to change the theme for the children form:
`{% form_theme bookingSealNumbers '...../sealNumberForm.html.twig' %}`

File: book/forms.rst (hunk 4 above)
Comment (weaverryan) [reply book/forms.rst]: This is all quite complex. Can you summarize this down to a note we should add? Or is it easier to say: "If you want to override a part of a form, go to the form profiler, find the field, and look at the block prefixes section". We could even add some screenshots. This seems maybe simpler than explaining the logic (unless the logic is quite simple - I'm arriving here late)

**Comment by juanpablomorenomartin** [comment]
This is the reason why Symfony has lost being in the top of the most used framework for the bad documentation it has, compared to other frameworks, without saying that the learning curve is vastly high, I thought that for the version 4 would improve in this aspect but it is still the same shit, now pa high you have to worry about the package that you will use to perform tasks as simple as a form and install it via composer in what shit Fabian would be thinking.

**Comment by javiereguiluz** [comment]
I'm closing this because we couldn't merge it on time and meanwhile, the docs about this "fragment naming" have been improved a lot. See https://symfony.com/doc/current/form/form_themes.html#form-fragment-naming Thank you all for the discussion.
